### ISSUE TITLE
Support ref readonly parameters

### ISSUE THREAD

**Issue by vzarytovskii** [dotnet/fsharp#16222]
Will fix https://github.com/dotnet/runtime/pull/89736

Should probably be as easy as just relaxing method resolution, since it's compile time only breakage (not a binary one, since we already support inrefs).

Ref readonly are encoded as inrefs ([in] &T) with special attribute: https://github.com/jjonescz/csharplang/blob/5134bd7042bcd71dbddbdff618b9b63dc54775a0/proposals/ref-readonly-parameters.md#metadata-encoding

More info: https://discord.com/channels/143867839282020352/312132327348240384/1169766448063856650

cc @Tarmil @Sergio0694 @TIHan

**Comment by teo-tsirpanis** [comment]
My understanding is that `inref<_>` is already equivalent to C#'s `ref readonly`. [You cannot pass an rvalue to an `inref<_>`](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AbEAzAzgHwxgBcACbACgRFIEsA7WbAHgeID4BKUgXlIs4BYAFAjspJEA==) for example.

If you want to fix `inref` and support passing rvalues to it, maybe the issue should be reworded accordingly.

**Comment by nodakai** [comment]
F# 8.0.100 sees a `ref readonly` parameter as a `byref` parameter, broke my code

```
Available overloads:
- Unsafe.AsRef<'T>(source: byref<'T>) : byref<'T> // Argument 'source' doesn't match
- Unsafe.AsRef<'T>(source: voidptr) : byref<'T> // Argument 'source' doesn't match
```

_C.f._ https://github.com/dotnet/runtime/issues/93251#issuecomment-1753544073

https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.asref?view=net-8.0#system-runtime-compilerservices-unsafe-asref-1(-0@)

**Comment by vzarytovskii** [comment]
> My understanding is that `inref<_>` is already equivalent to C#'s `ref readonly`. [You cannot pass an rvalue to an `inref<_>`](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AbEAzAzgHwxgBcACbACgRFIEsA7WbAHgeID4BKUgXlIs4BYAFAjspJEA==) for example.
> 
> 
> 
> If you want to fix `inref` and support passing rvalues to it, maybe the issue should be reworded accordingly.

Yeah. I actually think I'm going to close it now, since we have an ad-hoc fix which works, and it should be a small rfc if we want any additional support.

**Comment by vzarytovskii** [comment]
> F# 8.0.100 sees a `ref readonly` parameter as a `byref` parameter, broke my code
> 
> 
> 
> ```
> 
> Available overloads:
> 
> - Unsafe.AsRef<'T>(source: byref<'T>) : byref<'T> // Argument 'source' doesn't match
> 
> - Unsafe.AsRef<'T>(source: voidptr) : byref<'T> // Argument 'source' doesn't match
> 
> ```
> 
> 
> 
> _C.f._ https://github.com/dotnet/runtime/issues/93251#issuecomment-1753544073
> 
> 
> 
> https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.asref?view=net-8.0#system-runtime-compilerservices-unsafe-asref-1(-0@)

Yes, it's known. BCL changed the API and we didn't have much time to revert or fix on F# side. 8.0.101 should have the fix with will treat ref readonly as inrefs.

**Comment by nodakai** [comment]
> 8.0.101 should have the fix with will treat ref readonly as inrefs.

8.0.101 seems to be working fine for me. Thanks
